ble/FunctionPointerWithContext.h

```diff
@@ -107,19 +105,49 @@ class FunctionPointerWithContext {
 
 private:
     template<typename T>
-    static void membercaller(void *object, char *member, ContextType context) {
-        T *o = static_cast<T *>(object);
-        void (T::*m)(ContextType);
-        memcpy((char *)&m, member, sizeof(m));
-        (o->*m)(context);
+    static void membercaller(FunctionPointerWithContext* self, ContextType context) {
+        if(self->_memberFunctionAndPointer._object) {
+            T *o = static_cast<T *>(self->_memberFunctionAndPointer._object);        
+            void (T::*m)(ContextType);
+            memcpy((char*) &m, self->_memberFunctionAndPointer._memberFunction, sizeof(m));
+            (o->*m)(context);            
+        }
     }
 
-    void (*_function)(ContextType context);             /**< static function pointer - NULL if none attached */
-    void *_object;                                      /**< object this pointer - NULL if none attached */
-    char _member[16];                                   /**< raw member function pointer storage - converted back by
-                                                         *   registered _membercaller */
-    void (*_membercaller)(void *, char *, ContextType); /**< registered membercaller function to convert back and call
-                                                         *   _member on _object passing the context. */
+    static void functioncaller(FunctionPointerWithContext* self, ContextType context) {
+        if(self->_function) {
+            self->_function(context);
+        }
+    }
+
+    struct MemberFunctionAndPtr {
+        /*
+         * forward declaration of a class and a member function to this class. 
+         * Because the compiler doesnt know anything about the member function, it 
+         * will always take the biggest size that a member function can take.
+         * This also guarantee that the proper alignement will be chosen 
+         */
+        class UndefinedClass;
+        typedef void (UndefinedClass::*UndefinedMemberFunction)(ContextType);
+
+        void* _object;
+        union {
+            char _memberFunction[sizeof(UndefinedMemberFunction)];
+            UndefinedMemberFunction _allignement;            
+        };
+    };
+
+    union {
+        void (*_function)(ContextType context);             /**< static function pointer - NULL if none attached */        
+        /** 
+         * object this pointer and pointer to member - 
+         * _memberFunctionAndPointer._object will be NULL if none attached 
+         */        
+        MemberFunctionAndPtr _memberFunctionAndPointer;      
+    };
+
+    void (*_caller)(FunctionPointerWithContext*, ContextType);
```

Comment:
perhaps we could re-use `pFunctionPointerWithContext_t` instead of `FunctionPointerWithContext*` in `void (*_caller)(FunctionPointerWithContext*, ContextType);` ?